### PR TITLE
Bluetooth: Host: Free ACL RX fragments on disconnection

### LINKED ISSUE #76737: Bluetooth L2CAP TX fragmentation state not cleared upon disconnect
**Describe the bug**
When reconnecting, previous l2cap TX fragmentation state is kept, resulting in that we may send l2cap fragments to the controller without an l2cap header.

**To Reproduce**
* Run a connection with a long connection interval.
* Set max data length to 27 bytes.
* Send long ATT/L2CAP packets
* Disconnect one of the device
* Reconnect. Observe that the next L2CAP fragment will be sent without header

Setting `le_chan._pdu_remaining = 0` in [`bt_l2cap_chan_del()`](_pdu_remaining) resolves this issue.

**Expected behavior**
Upon disconnection all previous state shall be cleared. 

**Impact**
Showstopper

**Logs and console output**
Logs extracted from a custom test
```
bt_l2cap: L2CAP: push: pdu 0x819ddc8 len 200 cb 0x8055130 userdata (nil)
bt_l2cap: l2cap_pull: Adding L2CAP PDU header: buf 0x819ddf0 chan 0x8186280 len 200 / 200
app: disconnected: err 13 addr DC:52:5F:AA:9A:74 (random)

app: connected: err 0 addr DC:52:5F:AA:9A:74 (random)
bt_l2cap: L2CAP: push: pdu 0x819ddf0 len 200 cb 0x8055130 userdata (nil)
bt_l2cap: l2cap_pull: Not adding L2CAP header, handling remaining 123
```
**Environment (please complete the following information):**
Zephyr 3.7.0

**Additional context**

The bug was introduced in https://github.com/zephyrproject-rtos/zephyr/pull/72090

### LINKED ISSUE #76781: Bluetooth: Deadlock because of Missing L2CAP RX reassembly state cleanup on disconnect
**Describe the bug**
When reconnecting a Bluetooth connection, a deadlock may occur when the host issues an HCI command.
This occurs after that a link was disconnected before it received all L2CAP fragments of an L2CAP PDU.

**To Reproduce**
* Run a connection with a long connection interval.
* Set max data length to 27 bytes.
* Send long ATT/L2CAP packets
* Disconnect one of the devices
* Reconnect. Observe that the `bt_hci_cmd_send_sync()` triggers an HCI command timeout.

Calling `bt_conn_reset_rx_state(conn);` in `bt_conn_set_state()` when the old state is `BT_CONN_DISCONNECT_COMPLETE` and the new state is `BT_CONN_DISCONNECTED` resolves the issue.

**Expected behavior**
Upon disconnection all previous state shall be cleared.

**Impact**
Showstopper

**Logs and console output**
Logs extracted from a custom test
```
connected: err 0 addr F6:A4:15:3B:7A:F7 (random)
disconnected: err 16 addr F6:A4:15:3B:7A:F7 (random)
connected: err 0 addr F6:A4:15:3B:7A:F7 (random)
disconnected: err 16 addr F6:A4:15:3B:7A:F7 (random)
connected: err 0 addr F6:A4:15:3B:7A:F7 (random)
ASSERTION FAIL [err == 0] @ WEST_TOPDIR/zephyr/subsys/bluetooth/host/hci_core.c:411
  Controller unresponsive, command opcode 0x2016 timeout with err -11
  @ WEST_TOPDIR/zephyr/subsys/bluetooth/host/hci_core.c:411
  os: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
  os: Current thread: 0x8146760 (BT RX WQ)
```

**Environment (please complete the following information):**
Zephyr 3.7.0

**Additional context**
The bug was introduced in https://github.com/zephyrproject-rtos/zephyr/pull/72090

### PR DESCRIPTION
**Pull request by jori-nordic**
This function call frees the buffer kept by the host for reassembling L2CAP
PDUs into.

Without this call, the current buffer will eventually be
leaked, leading to a non-functional host due to lack of RX buffers.

The effect is worse when host flow control is not enabled, as the RX
buffer pool is shared with events, which means communication with the
controller is essentially dead.

Fixes #76781

### CODE REVIEW

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c [hunk 1]
```diff
@@ -0,0 +1,651 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
```
Comment (aescolar) [inline tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:2]: Note you have a few 2023 in the copyrights.

Comment (jori-nordic) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:2]: Thanks, will fix

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+		recv(buf);
+	}
+}
+
+static void le_read_buffer_size_complete(struct net_buf *rsp)
+{
+	struct bt_hci_rp_le_read_buffer_size *rp = (void *)rsp->data;
+
+	LOG_DBG("status 0x%02x", rp->status);
+	LOG_DBG("max len %d max num %d", rp->le_max_len, rp->le_max_num);
```
Comment (rugeGerritsen) [inline tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:340]: We should assert that `rp->le_max_len` is large enough to send a long l2cap packet. The other option is to rely on that the controller supports this length and validates when it receives it. If we do that we do not have to issue `BT_HCI_OP_LE_READ_BUFFER_SIZE`

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+	k_sem_take(&cmd_sem, K_FOREVER);
+	TEST_ASSERT(active_opcode == 0xFFFF, "");
+
+	active_opcode = opcode;
+
+	LOG_HEXDUMP_DBG(cmd->data, cmd->len, "HCI TX");
+	bt_send(cmd);
+
+	/* Wait until the command completes */
+	k_sem_take(&cmd_sem, K_FOREVER);
```
Comment (rugeGerritsen) [inline tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:303]: Why do we both give and take the `cmd_sem` here?

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c [hunk 4]
```diff
@@ -0,0 +1,650 @@
...
+
+		/* handlers should take a ref if they want to access the buffer
+		 * later
+		 */
+		net_buf_unref(buf);
+		return;
+	}
+
+	if (bt_buf_get_type(buf) == BT_BUF_ACL_IN) {
+		handle_acl(buf);
```
Comment (rugeGerritsen) [inline tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:277]: It looks like we are not doing anything with incoming data other than setting `flag_write_ack`. Is there something I've missed? If we are not using it, I would consider removing all the related code to simplify the test

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c [hunk 5]
```diff
@@ -0,0 +1,650 @@
...
+		TEST_FAIL("unhandled event 0x%x", hdr->evt);
+	}
+
+	LOG_DBG("opcode 0x%04x status %x", opcode, status);
+
+	TEST_ASSERT(status == 0x00, "cmd 0x%x status: 0x%x", opcode, status);
+
+	TEST_ASSERT(active_opcode == opcode, "unexpected opcode %x != %x", active_opcode, opcode);
+
+	if (active_opcode) {
```
Comment (rugeGerritsen) [inline tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:122]: Looks like `active_opcode` will always be true as we set it to `0xFFFF` or some other opcode. So I suggest removing this conditional statement.

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c
Comment (rugeGerritsen) [inline tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:1]: I guess there is some common code in here that can be useful for other tests in the future. Should that be extracted out to `bt_testlib_hci_host` or similar? 
No need to do that as part of this PR

File: tests/bsim/bluetooth/host/l2cap/reassembly/dut/src/dut.c [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+				     const void *data,
+				     uint16_t length)
+{
+	if (length) {
+		LOG_INF("RX notification");
+		LOG_HEXDUMP_DBG(data, length, "payload");
+
+		SET_FLAG(got_notification);
+	}
+
```
Comment (rugeGerritsen) [inline tests/bsim/bluetooth/host/l2cap/reassembly/dut/src/dut.c:42]: We should validate the content of the notification we got.

**Review by rugeGerritsen — CHANGES_REQUESTED** [review]
Thanks for adding a test for this bugfix.
Added some suggestions to improve the test

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c [hunk 7]
```diff
@@ -0,0 +1,650 @@
...
+	send_cmd(BT_HCI_OP_LE_WRITE_DEFAULT_DATA_LEN, buf, NULL);
+}
+
+static void set_data_len(void)
+{
+	uint16_t tx_octets, tx_time;
+
+	read_max_data_len(&tx_octets, &tx_time);
+	write_default_data_len(tx_octets, tx_time);
+}
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:379]: oh yeah. It's not used I think. I'll remove it.

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:1]: Yeah as you saw from the unused functions, it's copy-pasted from a bunch of other tests. I can make a lib if I have time later.

File: tests/bsim/bluetooth/host/l2cap/reassembly/dut/src/dut.c (hunk 6 above)
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/l2cap/reassembly/dut/src/dut.c:42]: should we? I guess I can, it makes the test stricter. 
The purpose was just checking for leaks though.

Comment (rugeGerritsen) [reply tests/bsim/bluetooth/host/l2cap/reassembly/dut/src/dut.c:42]: I have no strong opinion here. I just though it made sense to add some simple, cheap validation here.

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c [hunk 8]
```diff
@@ -0,0 +1,650 @@
...
+
+static void set_event_mask(uint16_t opcode)
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:381]: zephyr controller doesn't.

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c (hunk 2 above)
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:340]: what do you mean by long l2cap packet? isn't it decoupled from controller buffer size?

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c (hunk 3 above)
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:303]: had to scratch my head a bit haha. I'm re-using `cmd_sem` as a signal that we are able to send another command, which means that the current command (for which we took `cmd_sem` above) likely has gotten a response. 
We don't actually want to send anything more, so when we got that signal (ie the thread is un-suspended), then we release the sem immediately.

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c (hunk 4 above)
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:277]: Useful for debugging (ie if the remote starts sending unexpected ATT messages). I'll keep it.

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (rugeGerritsen) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:381]: FYI @cvinayak , the spec states:
> The value with all bits set to 0 indicates that no events are specified. The
default is for bits 0 to 4 (the value 0x0000 0000 0000 001F) to be set.

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c (hunk 2 above)
Comment (rugeGerritsen) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:340]: I meant that the buffer size is large enough to fit an L2CAP packet in a single HCI Data packet right

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c (hunk 3 above)
Comment (rugeGerritsen) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:303]: Maybe add some docs here to explain that?

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c (hunk 4 above)
Comment (rugeGerritsen) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:277]: Please remove `flag_write_ack` then

File: tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+	static uint8_t data[] = NOTIFICATION_PAYLOAD;
+	uint16_t handle = GATT_HANDLE;
+	struct net_buf *att_packet = alloc_l2cap_pdu();
+
+	/* Prepare (G)ATT notification packet */
+	net_buf_add_u8(att_packet, BT_ATT_OP_NOTIFY);
+	net_buf_add_le16(att_packet, handle);
+	net_buf_add_mem(att_packet, data, sizeof(data));
+
+	size_t on_air_size = 5;
```
Comment (rugeGerritsen) [inline tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:545]: Where does the `on_air` value 5 come from? 
I would prefer that NOTIFICATION_PAYLOAD is longer than 27 bytes to guarantee that the controller cannot send the data in a single packet as some implementations may choose to buffer payload until nothing more can fit.

Comment (rugeGerritsen) [reply tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c:545]: That makes sense!

**Review by rugeGerritsen — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Review by ubieda — APPROVED** [review]
LGTM - Nice branch-name BTW

**Comment by jhedberg** [comment]
@jori-nordic shouldn't this also be marked as fixing #76737?

**Comment by jori-nordic** [comment]
No, I messed up when I tagged the issue initially.
The fix for TX is still being worked on (the test rather). I'll submit it next week.

**Pull request merged**